Sources/TiqrCoreObjC/Classes/AuthenticationChallenge.m

```diff
@@ -148,5 +143,102 @@ + (AuthenticationChallenge *)challengeWithChallengeString:(NSString *)challengeS
+    
+    NSURLComponents * components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
+    
+    // Parse the parameters
+    NSString *serverIdentifier = [self getQueryParameter:@"i" fromComponents:components];
+    // User is optional
+    NSString *user = [self getQueryParameter:@"u" fromComponents:components];
+
+    if (!serverIdentifier) {
+        NSLog(@"Required parameter 'i' missing from the authentication URL!");
+        [self generateInvalidQRCodeError: error];
+        return nil;
+    }
+    
+    if(![self findIdentityForServerIdentifier: serverIdentifier andUser:user forChallenge:challenge error: error]) {
+        return nil;
+    }
+    
+    NSString *sessionKey = [self getQueryParameter:@"s" fromComponents:components];
+    if (!sessionKey) {
+        NSLog(@"Required parameter 's' missing from the authentication URL!");
+        [self generateInvalidQRCodeError: error];
+        return nil;
+    }
+    NSString *challengeParam = [self getQueryParameter:@"c" fromComponents:components];
+    if (!challengeParam) {
+        NSLog(@"Required parameter 'c' missing from the authentication URL!");
+        [self generateInvalidQRCodeError: error];
+        return nil;
+    }
+    
+    challenge.sessionKey = sessionKey;
+    challenge.challenge = challengeParam;
+    challenge.serviceProviderDisplayName = challenge.identityProvider.displayName;
+    challenge.serviceProviderIdentifier = @"";
+    
+    NSString *protocolVersion = [self getQueryParameter:@"v" fromComponents:components];
+    if (!protocolVersion) {
+        protocolVersion = @"1";
+    }
+    challenge.protocolVersion = protocolVersion;
+    challenge.returnUrl = nil;
+    return challenge;
+
+}
+
++(BOOL) findIdentityForServerIdentifier:(NSString * _Nonnull)serverIdentifier andUser:(NSString * _Nullable)user forChallenge:(AuthenticationChallenge *)challenge error:(NSError **)error {
+    IdentityService *identityService = ServiceContainer.sharedInstance.identityService;
+    IdentityProvider *identityProvider = [identityService findIdentityProviderWithIdentifier:serverIdentifier];
+    if (identityProvider == nil) {
+        [self generateUnknownIdentityError:error];
+        return NO;
+    }
+    if (user) {
+        Identity *identity = [identityService findIdentityWithIdentifier:user forIdentityProvider:identityProvider];
+        if (identity == nil) {
+            [self generateInvalidAccountError:error];
+            return NO;
+        }
+        
+        challenge.identities = @[identity];
+        challenge.identity = identity;
+    } else {
+        NSArray *identities = [identityService findIdentitiesForIdentityProvider:identityProvider];
+        if (identities == nil || [identities count] == 0) {
+            [self generateInvalidAccountError:error];
+            return NO;
+        }
+        
+        challenge.identities = identities;
+        challenge.identity = [identities count] == 1 ? identities[0] : nil;
```

Comment:
`identities.firstObject`

Reply:
This is not the same as `firstObject`, because we do a hard check on `count == 1`